crates/ra_hir_def/src/nameres/collector.rs

```diff
@@ -36,10 +37,13 @@ pub(super) fn collect_defs(db: &impl DefDatabase2, mut def_map: CrateDefMap) ->
         );
 
         // look for the prelude
-        if def_map.prelude.is_none() {
+        // If the prelude is the "core" prelude, try to replace it with a higher
+        // level prelude (e.g. "std") if available.
+        if def_map.prelude.is_none() || prelude_is_core {
             let map = db.crate_def_map(dep.crate_id);
             if map.prelude.is_some() {
                 def_map.prelude = map.prelude;
+                prelude_is_core = dep.name == "core";
```

Comment:
Let me know if this is the right thing to do, or whether a string comparison is not good enough. As far as I can see the crate id for libcore is not available in this method. But I think since this compares crate names and those have to be unique it should be ok?